src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Pkcs12SafeBag.cs

```diff
@@ -0,0 +1,143 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography.Asn1;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    public abstract class Pkcs12SafeBag
+    {
+        private readonly string _bagIdValue;
+        private Oid _bagOid;
+        private CryptographicAttributeObjectCollection _attributes;
+
+        public ReadOnlyMemory<byte> EncodedBagValue { get; }
+
+        public CryptographicAttributeObjectCollection Attributes
+        {
+            get
+            {
+                if (_attributes == null)
+                {
+                    _attributes = new CryptographicAttributeObjectCollection();
+                }
+
+                return _attributes;
+            }
+
+            internal set
+            {
+                Debug.Assert(value != null);
+                _attributes = value;
+            }
+        }
+
+        protected Pkcs12SafeBag(string bagIdValue, ReadOnlyMemory<byte> encodedBagValue, bool skipCopy=false)
+        {
+            if (string.IsNullOrEmpty(bagIdValue))
+                throw new ArgumentNullException(nameof(bagIdValue));
+
+            // Read to ensure that there is precisely one legally encoded value.
+            AsnReader reader = new AsnReader(encodedBagValue, AsnEncodingRules.BER);
+            reader.GetEncodedValue();
+            reader.ThrowIfNotEmpty();
+
+            _bagIdValue = bagIdValue;
+            EncodedBagValue = skipCopy ? encodedBagValue : encodedBagValue.ToArray();
+        }
+
+        public byte[] Encode()
+        {
+            using (AsnWriter writer = EncodeToNewWriter())
+            {
+                return writer.Encode();
+            }
+        }
+
+        public Oid GetBagId()
+        {
+            if (_bagOid == null)
+            {
+                _bagOid = new Oid(_bagIdValue);
+            }
+
+            return new Oid(_bagOid);
+        }
```

Comment:
Why is this creating one Oid and then cloning it?  It's creating the first one to cache it to avoid multiple parsings of _bagIdValue, but then cloning to avoid potential mutation of the cached instance?

Reply:
Yep